File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/Transfer.java
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2012 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */

package edu.berkeley.boinc.rpc;

import java.io.Serializable;

import android.os.Parcel;
import android.os.Parcelable;

public class Transfer implements Serializable, Parcelable {
    private static final long serialVersionUID = 1L;
    public String name;
    public String project_url;
    public boolean generated_locally;
    public long nbytes;
    public boolean xfer_active;
    public boolean is_upload;
    public int status;
    public long next_request_time;
    public long time_so_far;
    public long bytes_xferred;
    public float xfer_speed;
    public long project_backoff;

    @Override
    public int describeContents() {
        // TODO Auto-generated method stub
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        // TODO Auto-generated method stub
        dest.writeString(name);
        dest.writeString(project_url);
        dest.writeLong(nbytes);
        dest.writeInt(status);
        dest.writeLong(next_request_time);
        dest.writeLong(time_so_far);
        dest.writeLong(bytes_xferred);
        dest.writeFloat(xfer_speed);
        dest.writeLong(project_backoff);

        dest.writeBooleanArray(new boolean[]{
                generated_locally,
                xfer_active,
                is_upload
        });
    }

    public Transfer() {
    }

    private Transfer(Parcel in) {
        name = in.readString();
        project_url = in.readString();
        nbytes = in.readLong();
        status = in.readInt();
        next_request_time = in.readLong();
        time_so_far = in.readLong();
        bytes_xferred = in.readLong();
        xfer_speed = in.readFloat();
        project_backoff = in.readLong();

        boolean[] bArray = in.createBooleanArray();
        generated_locally = bArray[0];
        xfer_active = bArray[1];
        is_upload = bArray[2];
    }

    public static final Parcelable.Creator<Transfer> CREATOR = new Parcelable.Creator<Transfer>() {
        public Transfer createFromParcel(Parcel in) {
            return new Transfer(in);
        }

        public Transfer[] newArray(int size) {
            return new Transfer[size];
        }
    };
}

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/Transfer.kt
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2020 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc.rpc

import android.os.Parcel
import android.os.Parcelable
import androidx.core.os.ParcelCompat.readBoolean
import androidx.core.os.ParcelCompat.writeBoolean
import java.io.Serializable

enum class TransferStatus(val status: Int) {
    ERR_GIVEUP_DOWNLOAD(-114),
    ERR_GIVEUP_UPLOAD(-115)
}

data class Transfer(
        var name: String = "",
        var projectUrl: String = "",
        var noOfBytes: Long = 0,
        var status: Int = 0,
        var nextRequestTime: Long = 0,
        var timeSoFar: Long = 0,
        var bytesTransferred: Long = 0,
        var transferSpeed: Float = 0f,
        var projectBackoff: Long = 0,
        var generatedLocally: Boolean = false,
        var isTransferActive: Boolean = false,
        var isUpload: Boolean = false
) : Serializable, Parcelable {
    private constructor (parcel: Parcel) :
            this(parcel.readString() ?: "", parcel.readString() ?: "",
                    parcel.readLong(), parcel.readInt(), parcel.readLong(), parcel.readLong(),
                    parcel.readLong(), parcel.readFloat(), parcel.readLong(), readBoolean(parcel),
                    readBoolean(parcel), readBoolean(parcel))

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeString(projectUrl)
        dest.writeLong(noOfBytes)
        dest.writeInt(status)
        dest.writeLong(nextRequestTime)
        dest.writeLong(timeSoFar)
        dest.writeLong(bytesTransferred)
        dest.writeFloat(transferSpeed)
        dest.writeLong(projectBackoff)
        writeBoolean(dest, generatedLocally)
        writeBoolean(dest, isTransferActive)
        writeBoolean(dest, isUpload)
    }

    object Fields {
        const val GENERATED_LOCALLY = "generated_locally"
        const val NBYTES = "nbytes"
        const val IS_UPLOAD = "is_upload"
        const val STATUS = "status"
        const val NEXT_REQUEST_TIME = "next_request_time"
        const val TIME_SO_FAR = "time_so_far"
        const val BYTES_XFERRED = "bytes_xferred"
        const val XFER_SPEED = "xfer_speed"
        const val PROJECT_BACKOFF = "project_backoff"
    }

    companion object {
        private const val serialVersionUID = 1L

        @JvmField
        val CREATOR: Parcelable.Creator<Transfer> = object : Parcelable.Creator<Transfer> {
            override fun createFromParcel(parcel: Parcel) = Transfer(parcel)

            override fun newArray(size: Int) = arrayOfNulls<Transfer>(size)
        }
    }
}
